Remove `:feature` checks on some EE code

johnswanson opened the pull request:
<!-- Added by 'Add Issue References to PR' GitHub Action. To disable linking, add 'no-auto-issue-links' label to your PR. --> closes #61357

johnswanson committed:
Assert premium features for some middleware

johnswanson committed:
Fix an existing bug

These tests started failing here, which revealed that they formerly
never actually were testing what they thought they were.

A few reasons:
- they turned on `advanced-permissions` feature by itself, so sandboxing
was always off here

- they set up sandboxing in a nonstandard way

- one had a typo so we weren't even wrapping the body with
`with-download-perms!` (it was a separate block entirely)

This hid the bug that some of the tests were intended to reproduce - if
a card used for sandboxing had a native query, then we required that you
have download permissions on the entire database, which is wrong.

johnswanson committed:
fix test

johnswanson committed:
make assertions with/without tokens for sandboxing

camsaul commented on enterprise/backend/src/metabase_enterprise/database_routing/middleware.clj:
```diff
@@ -33,14 +34,15 @@
 (defenterprise attach-destination-db-middleware
   "Pre-processing middleware. Calculates the destination database that should be used for this query, e.g. for caching
   purposes. Does not make any changes to the query besides (possibly) adding a `:destination-database/id` key."
-  :feature :database-routing
```
You can't be removing token feature checks and then not at least putting a note about why it doesn't do the feature check. Otherwise someone else is just going to look at this code that looks wrong and fix it

johnswanson committed:
Add comments about why `:feature :none`

camsaul commented on enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/sandboxing_test.clj:
```diff
@@ -389,7 +419,17 @@
+                                     (qp/process-query {:database (mt/id)
+                                                        :type :query
+                                                        :query {:source-table (mt/id :venues)
+                                                                :limit 1}}))))))))))))))))
```
You can use this, it's much more concise.

```suggestion
                                     (qp/process-query (mt/mbql-query venues {:limit 1})}))))))))))))))))
```

johnswanson committed:
Update enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/sandboxing_test.clj

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

johnswanson committed:
add a nicer docstring and change fn name

replace `any-native-unsandboxed-stage?` with
`any-native-stage-not-introduced-by-sandbox?`

dpsutton commented on enterprise/backend/src/metabase_enterprise/database_routing/middleware.clj:
```diff
@@ -33,14 +34,17 @@
   purposes. Does not make any changes to the query besides (possibly) adding a `:destination-database/id` key."
-  :feature :database-routing
+  ;; run this even when the `:database-routing` feature is not enabled, so that we can assert that it *is* enabled if
+  ;; DB routing is configured. (Throwing here is better than silently ignoring the configured routing.)
```
It's worth calling out a bit of a subtlety. We want to only throw in these features when the feature _is configured_. Lots of these things are in the style of "apply sandboxing if configured", and these get called on every query. It's only when we see something like "sandbox table 4 on field 17 = 'gadget'" that we want to assert that sandboxing is enabled, and _fail closed_ otherwise.

camsaul reviewed (approved):
I've read thru this completely twice and AFAIK this LGTM. I think we should revisit the `defenterprise` stuff and come up with a way to have separate EE implementations if you have an appropriate token or don't (discussed in Slack), but that can happen in a separate PR.

johnswanson committed:
Block without feature enabled

johnswanson committed:
CR response
